src/test/kotlin/com/lykke/trade/volume/monitoring/service/process/AssetVolumeConverterTest.kt

```diff
@@ -47,33 +46,87 @@ class AssetVolumeConverterTest {
+        assertEquals(BigDecimal("1.56"), converter.convert("Asset1", BigDecimal.ONE, emptySet(), "Asset2"))
+        assertEquals(BigDecimal("4.86"), converter.convert("Asset1", BigDecimal("3.1234567"), emptySet(), "Asset2"))
     }
 
     @Test
     fun testConvertQuotingToBase() {
-        assertEquals(BigDecimal("1.0029"), converter.convert("Asset2", BigDecimal("1.56"), "Asset1"))
-        assertEquals(BigDecimal("3.1243"), converter.convert("Asset2", BigDecimal("4.86"), "Asset1"))
+        assertEquals(BigDecimal("1.0029"), converter.convert("Asset2", BigDecimal("1.56"), emptySet(), "Asset1"))
+        assertEquals(BigDecimal("3.1243"), converter.convert("Asset2", BigDecimal("4.86"), emptySet(), "Asset1"))
     }
 
     @Test
     fun testConvertToSameAsset() {
-        assertEquals(BigDecimal.ONE, converter.convert("Asset1", BigDecimal.ONE, "Asset1"))
+        assertEquals(BigDecimal.ONE, converter.convert("Asset1", BigDecimal.ONE, emptySet(), "Asset1"))
     }
 
     @Test(expected = ApplicationException::class)
     fun testUnknownTargetAsset() {
-        converter.convert("Asset1", BigDecimal.ONE, "UnknownAsset")
+        converter.convert("Asset1", BigDecimal.ONE, emptySet(), "UnknownAsset")
     }
 
     @Test(expected = ApplicationException::class)
     fun testUnknownAssetPair() {
-        converter.convert("Asset3", BigDecimal.ONE, "Asset2")
+        converter.convert("Asset3", BigDecimal.ONE, emptySet(), "Asset2")
+    }
+
+    @Test(expected = ApplicationException::class)
+    fun testUnknownStraightAndCrossAssetPairs() {
+        converter.convert("Asset3", BigDecimal.ONE, setOf("CrossAsset1", "CrossAsset2", "CrossAsset3"), "Asset2")
+    }
+
+    @Test
+    fun testConvertUnknownAssetPairThroughCrossAsset() {
+        Mockito.`when`(assetPairsHolder.getAssetPair("Asset3", "CrossAsset2"))
+                .thenReturn(AssetPair("CrossPair2-1", "Asset3", "CrossAsset2", 1))
+        Mockito.`when`(assetPairsHolder.getAssetPair("CrossAsset2", "Asset2"))
+                .thenReturn(AssetPair("CrossPair2-2-WithoutPrice", "Asset2", "CrossAsset2", 2))
+
+        Mockito.`when`(assetPairsHolder.getAssetPair("Asset3", "CrossAsset3"))
+                .thenReturn(AssetPair("CrossPair3-1", "Asset3", "CrossAsset3", 3))
+        Mockito.`when`(assetPairsHolder.getAssetPair("CrossAsset3", "Asset2"))
+                .thenReturn(AssetPair("CrossPair3-2", "Asset2", "CrossAsset3", 4))
+
+        Mockito.`when`(pricesHolder.getPrice("CrossPair2-1"))
+                .thenReturn(BigDecimal("1.1"))
+        Mockito.`when`(pricesHolder.getPrice("CrossPair3-1"))
+                .thenReturn(BigDecimal("10.123"))
+        Mockito.`when`(pricesHolder.getPrice("CrossPair3-2"))
+                .thenReturn(BigDecimal("15.6789"))
+
+        val result = converter.convert("Asset3", BigDecimal("1.23"), setOf("CrossAsset1", "CrossAsset2", "CrossAsset3"), "Asset2")
```

Comment:
suggest using `sortedSet` to preserve input order

Reply:
made config value `crossAssetIds` a list as order is important